### PR TITLE
Llin/alt offer usage emails

### PR DESCRIPTION
**Pull request by long74100**
# ⛔️ DEPRECATION WARNING

**This repository is deprecated and in maintainence-only operation while we work on a replacement, please see [this announcement](https://discuss.openedx.org/t/deprecation-removal-ecommerce-service-depr-22/6839) for more information.**

Although we have stopped integrating new contributions, we always appreciate security disclosures and patches sent to [security@edx.org](mailto:security@edx.org)

## Required Testing
- [ ] Before deploying this change, complete a purchase in the stage environment. 
(^ We can remove that manual check once REV-2624 is done and the corresponding e2e test runs again)

## Description

Describe what this pull request changes, and why these changes were made. How will these changes affect other people, installations of edx, etc.?
Please include links to any relevant ADRs, design artifacts, and decision documents. Make sure to document the rationale behind significant changes in the repo, per [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:
- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author", "Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change; how did YOU test this change?

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, OpenEdx vs. edx.org differences, development vs. production environment differences, security, or accessibility.

### CODE REVIEW

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 1]
```diff
@@ -72,26 +61,150 @@ def get_booking_limits(site, offer):
...
+        """
+        last_email_of_type_sent = OfferUsageEmail.objects.filter(offer=enterprise_offer, email_type=email_type).last()
+
+        if not last_email_of_type_sent:
+            return True
+
+        email_metadata = last_email_of_type_sent.offer_email_metadata
+
+        # All emails sent after 6/28/22 should have this field in the metadata
+        previous_usage_data = email_metadata.get('email_usage_data', {})
```
Comment (christopappas) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: nit: i think the default here could be left off. `.get` returns None and should be caught by line 92 if not found

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 2]
```diff
@@ -72,26 +61,150 @@ def get_booking_limits(site, offer):
...
+            OfferUsageEmailTypes.OUT_OF_BALANCE,
+            total_limit
+        )
+
+        if not should_send_email:
+            return (False, EMAIL_SENT_BEFORE_OFFER_NOT_REUPPED)
+
+        if is_enrollment_limit_offer:
+            return (percentage_usage == 100, THRESHOLD_NOT_REACHED)
+        else:
```
Comment (christopappas) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: nit: don't need the else

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 3]
```diff
@@ -72,26 +61,150 @@ def get_booking_limits(site, offer):
...
+        enterprise_offer,
+        usage_info,
+    ):
+        """
+        Return whether an offer is eligible for the low balance email.
+        """
+        percentage_usage = usage_info['percentage_usage']
+        total_limit = usage_info['total_limit']
+
+        return percentage_usage >= 75 and self.should_send_email_type(enterprise_offer, OfferUsageEmailTypes.LOW_BALANCE, total_limit)
```
Comment (christopappas) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: nit: it's personal preference, but perhaps splitting this out on a few lines to make the return line a bit more concise

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 4]
```diff
@@ -23,39 +24,27 @@
 
 EMAIL_SUBJECT = 'Offer Usage Notification'
 
+# Reasons why an email should not be sent
+THRESHOLD_NOT_REACHED = 'Threshold not reached.'
+EMAIL_SENT_BEFORE_OFFER_NOT_REUPPED = 'Email sent before, offer has not been reupped.'
```
Comment (iloveagent57) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: nit: re-upped is not a super common term, maybe something about "balance replenished" instead?

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 5]
```diff
@@ -139,18 +260,23 @@ def handle(self, *args, **options):
...
                 task_result.get(propagate=False)
                 if task_result.successful():
                     successful_send_count += 1
-                    OfferUsageEmail.create_record(enterprise_offer, meta_data={
-                        'email_usage_data': email_body_variables,
-                        'email_subject': EMAIL_SUBJECT,
-                        'email_addresses': enterprise_offer.emails_for_usage_alert
-                    })
+                    OfferUsageEmail.create_record(
+                        email_type=OfferUsageEmailTypes.DIGEST,
```
Comment (iloveagent57) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: do you want `email_type=email_type` here?

**Review by iloveagent57 — COMMENTED** [review]
Looks great, really just the one question about persisting email_type as digest.
I don't feel strongly about "reupped" vs. "balance replenished" (although maybe spelling it as "re-upped" would make it clearer?).

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 6]
```diff
@@ -72,26 +61,150 @@ def get_booking_limits(site, offer):
...
+            total_limit
+        )
+
+        if not should_send_email:
+            return (False, EMAIL_SENT_BEFORE_OFFER_NOT_REUPPED)
+
+        if is_enrollment_limit_offer:
+            return (percentage_usage == 100, THRESHOLD_NOT_REACHED)
+        else:
+            return (total_limit - current_usage <= 100, THRESHOLD_NOT_REACHED)
```
Comment (christopappas) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: [question] is the `100` here a percentage? (If so, would this always return True because all %'s <= 100?)

File: ecommerce/enterprise/tests/test_send_enterprise_offer_limit_emails.py [hunk 7]
```diff
@@ -36,130 +38,314 @@ def setUp(self):
...
+            json={
+                'max_discount': max_discount,
+                'percent_of_offer_spent': (amount_of_offer_spent / max_discount) * 100,
+                'amount_of_offer_spent': amount_of_offer_spent,
+            },
+            content_type='application/json',
+        )
+
+    @responses.activate
+    def test_low_balance_email(self):
```
Comment (christopappas) [inline ecommerce/enterprise/tests/test_send_enterprise_offer_limit_emails.py:72]: good testing the re-upped bit. 👍  i wonder if it's worth adding a test/assertion to verify the negative (emails not getting sent when they shouldn't).

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py (hunk 5 above)
Comment (long74100) [reply ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: 🤦 absolutely

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py (hunk 1 above)
Comment (long74100) [reply ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: True

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py (hunk 6 above)
Comment (long74100) [reply ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: the 100 here is the dollar value amount. If remaining_balance <= 100, we want to set this email.

File: ecommerce/settings/base.py [hunk 8]
```diff
@@ -853,3 +854,12 @@
 ENTERPRISE_EMAIL_FILE_ATTACHMENTS_BUCKET_NAME = ''
 ENTERPRISE_EMAIL_FILE_ATTACHMENTS_BUCKET_LOCATION = 'us-east-1'  # change this when developing with your own bucket
 
+BRAZE_OFFER_DIGEST_CAMPAIGN = ''
+BRAZE_OFFER_LOW_BALANCE_CAMPAIGN = ''
+BRAZE_NO_BALANCE_CAMPAIGN = ''
```
Comment (christopappas) [inline ecommerce/settings/base.py]: nit: should this be: `BRAZE_OFFER_NO_BALANCE_CAMPAIGN`

File: ecommerce/enterprise/tests/test_send_enterprise_offer_limit_emails.py (hunk 7 above)
Comment (long74100) [reply ecommerce/enterprise/tests/test_send_enterprise_offer_limit_emails.py:72]: I think this is covered by the 3 different tests. Emails shouldn't be sent when no balance email has been sent, or when the digest email has been sent recently.

File: ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py [hunk 9]
```diff
@@ -72,26 +61,150 @@ def get_booking_limits(site, offer):
...
+        email_type = self.get_email_type(offer, usage_info, is_enrollment_limit_offer)
+
         return {
+            'email_type': email_type,
             'percent_usage': percentage_usage,
             'total_limit': total_limit if is_enrollment_limit_offer else "${}".format(total_limit),
+            'total_limit_raw': total_limit,
             'offer_type': 'Enrollment' if is_enrollment_limit_offer else 'Booking',
             'offer_name': offer.name,
             'current_usage': current_usage if is_enrollment_limit_offer else "${}".format(current_usage),
```
Comment (iloveagent57) [inline ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: thinking out loud: should we add `is_enrollment_limit_offer` as a key to this dictionary?  Then we can have templates or downstream frontends conditionally format limits and usage however they want based on that boolean?

Comment (iloveagent57) [reply ecommerce/enterprise/management/commands/send_enterprise_offer_limit_emails.py]: ...and if we do that, I'd suggest changing `total_limit` to `total_limit_str` and `total_limit_raw` to just `total_limit`.  Ditto for `current_usage`.

**Review by iloveagent57 — APPROVED** [review]
Minor suggestion to give us more flexibility on the template side of things.
